avalon/pipeline.py

```diff
@@ -926,61 +926,111 @@ def get_representation_context(representation):
     return context
 
 
-def update_current_task(task=None, asset=None, app=None):
-    """Update active Session to a new task work area.
+def compute_session_changes(session, task=None, asset=None, app=None):
+    """Compute the changes for a Session object on asset, task or app switch
 
-    This updates the live Session to a different `asset`, `task` or `app`.
+    This does *NOT* update the Session object, but returns the changes
+    required for a valid update of the Session.
 
     Args:
-        task (str): The task to set.
-        asset (str): The asset to set.
-        app (str): The app to set.
+        session (dict): The initial session to compute changes to.
+            This is required for computing the full Work Directory, as that
+            also depends on the values that haven't changed.
+        task (str, Optional): Name of task to switch to.
+        asset (str or dict, Optional): Name of asset to switch to.
+            You can also directly provide the Asset dictionary as returned
+            from the database to avoid an additional query. (optimization)
+        app (str, Optional): Name of app to switch to.
 
     Returns:
-        dict: The changed key, values in the current Session.
+        dict: The required changes in the Session dictionary.
 
     """
 
+    changes = dict()
+
+    # If no changes, return directly
+    if not any([task, asset, app]):
+        return changes
+
+    # Get asset document and asset
+    asset_document = None
+    if asset:
+        if isinstance(asset, dict):
+            # Assume asset database document
+            asset_document = asset
+            asset = asset["name"]
+        else:
+            # Assume asset name
+            asset_document = io.find_one({"name": asset,
+                                          "type": "asset"})
+            assert asset_document, "Asset must exist"
+
+    # Detect any changes compared session
     mapping = {
         "AVALON_ASSET": asset,
         "AVALON_TASK": task,
         "AVALON_APP": app,
     }
-    changed = {key: value for key, value in mapping.items() if value}
-    if not changed:
-        return
+    changes = {key: value for key, value in mapping.items() if value
+               and value != session.get(key)}
```

Comment:
This is trivial, but my IDE is highlighting this line because of PEP8 (line break before binary operator).

Reply:
So would it then be like this? 
```python
changes = {key: value for key, value in mapping.items() 
           if value and value != session.get(key)}
```

Reply:
Yes, that would work !

Reply:
Fixed with https://github.com/getavalon/core/pull/442/commits/3d7c65e1c3aecc92fa4a0a08a174af2c940df8de